DM-29266: Add translator methods to read all relevant HDUs

timj opened the pull request.

timj committed:
Clean up some defaults and imports

timj committed:
Add translator method for reading all headers from a file

timj committed:
Change astrometadata translate to use native header reading

DECam and CFHT now report all HDUs.

timj committed:
Remove file that was added by mistake.

timj committed:
Improve some error messages by including the filename

timj committed:
If the file is not readable but we have primary return it

Sometimes we have read the header from a YAML file and in that
scenario DECam and CFHT code should not try to treat it as
a raw fits file.

timj committed:
Adjust test for raise in different place

The code now asks for the metadata translator class earlier.

timj committed:
Only merge by default if hdu num is negative

HSC compressed files have two HDUs but uncompressed only
have one. The metadata is all in the second so we can't avoid
doing the merge. Only attempt it though if EXTEND=T.

timj committed:
For default translator reader, merge primary and secondary headers

If something more complicated is required the translator class
should implement a specialist reader.

kfindeisen commented on python/astro_metadata_translator/file_helpers.py:
```diff
@@ -111,7 +112,7 @@ def find_files(files, regex):
     return found_files
 
 
-def read_basic_metadata_from_file(file, hdrnum, errstream=sys.stderr, can_raise=False):
+def read_basic_metadata_from_file(file, hdrnum, errstream=sys.stderr, can_raise=True):
```
Does this change break any code that assumes this function returns `None`?

timj replied:
No. It's only used internally by this module for now (it was only added a couple of weeks back) and I was making it consistent with the docstring. Most callers have been specifying it explicitly.

kfindeisen commented on python/astro_metadata_translator/translator.py:
```diff
@@ -943,6 +943,39 @@ def to_observation_counter(self):
+        """
+        if primary is not None:
+            yield primary
+        else:
```
This causes `read_all_headers` to return only the primary header if the `primary` keyword is set. Is this intended? It contradicts the docstring and method name.

timj replied:
That is intended. In the general case if the routine is given the primary header it's unlikely that this routine can guess well enough to improve on the situation. I was trying to make it simple. Would you prefer that I look for an INHERIT=T header and EXTEND=T header and by default try to read the second header and merge it? I don't think it can do anything in the default case other than return one header (merged or not) since for an arbitrary FITS file of unknown origin I can't tell if subsequent HDUs are (say) amplifiers or detector images.

timj replied:
The other point here is that I don't want to be forced into reading every header twice. I have to read it once in order to work out which metadata translator class to call. In many cases that header that I read is sufficient to do metadata translation so it's much better to return it immediately rather than require that we then read the same header a second time (or possibly third time if we merged in the caller because it was a compressed image like those in testdata_ci_hsc).

kfindeisen replied:
Actually, what I had in mind was not having a default implementation at all, but then you'd need a `SingleHeaderTranslator` subclass that everything other than LSST/DECam/CHFT use.

If you keep the code organized as-is, I recommend at least including a note in the docstring to the effect of:
> This implementation returns `primary`, if provided, otherwise it reads the first header from `filename`. Translator subclasses for instruments with additional headers must override this method.

Otherwise, subclass authors have no way of knowing that supporting multiple headers is not the base class's responsibility.

timj replied:
Okay. I really don't want to have to insert a new class into the hierarchy that will be the default so I will make it clearer in the documentation what the default implementation will do.

timj replied:
`read_translatable_heaers` ?

timj replied:
I have completely revamped the documentation and changed the method name to `determine_translatable_headers`

kfindeisen commented on python/astro_metadata_translator/translators/decam.py:
```diff
@@ -297,3 +298,57 @@ def fix_header(cls, header, instrument, obsid, filename=None):
+            without already knowing what data is in the file. For many
+            instruments where the primary header is the only relevant
+            header, the primary header will be returned with no further
+            action.
```
Consider rewording this to be more DECam-specific.

kfindeisen commented on python/astro_metadata_translator/translator.py:
```diff
@@ -943,6 +943,39 @@ def to_observation_counter(self):
+            yield primary
+        else:
+            # Prevent circular import by deferring
+            from .file_helpers import read_basic_metadata_from_file
```
Is the circular dependency between `.headers` and `MetadataTranslator` really necessary? `merge_headers` seems to already depend on `MetadataTranslator`, but does this code need to be part of `MetadataTranslator`?

timj replied:
How would I read the file header if I removed it? Are you saying that I should write the header reading explicitly with astropy.io.fits as I do in the subclasses? I don't think that works if I end up needing to use merge_headers.

I think there are two paths for circularity. One is that file_helpers imports ObservationInfo which imports MetadataTranslator which would try to import file_helpers. The other is that file_helpers imports merge_headers which imports MetadataTranslator to sort the headers by observing time.

kfindeisen replied:
Ok, I missed the `ObservationInfo` angle. I think the fact that it's so easy to get circular dependencies is a sign of more fundamental design problems -- perhaps that `MetadataTranslator` needs to be broken up into smaller, more self-contained pieces. However, that does sound out of scope for this ticket.

timj replied:
It's the base class for all translators so it's difficult. The problem is that the code that works out whether a translator can be used for a specific header comes from the registered translator subclasses (which all use init_subclass to do the registration automatically). Therefore any helper code that could benefit from looking at a header and working out the relevant translator must call back into MetadataTranslator.  This didn't matter until MetadataTranslator started having to read file headers.

kfindeisen commented on python/astro_metadata_translator/translators/decam.py:
```diff
@@ -297,3 +298,57 @@ def fix_header(cls, header, instrument, obsid, filename=None):
+                    continue
+
+                header = hdu.header
+                if "CCDNUM" not in header:
```
Can you comment on when this is expected to happen?

timj replied:
I assume it's because the primary header doesn't have a CCDNUM and it will skip it. Maybe there are other HDUs that should be skipped as well. I didn't write the code when it was in obs_decam. Maybe it was @parejkoj ?

parejkoj replied:
Can you please point me to the obs_decam lines you're referring to? I need some context here (if it even was me).

timj replied:
It was added in 1f9b698

timj replied:
Sorry. The SHA1 won't work from here.  Try this: https://github.com/lsst/obs_decam/commit/1f9b698deb4976870652e993550d77935ea47b67

timj replied:
and I've just realized you were asking about the CCDNUM test and not the CCDNUM usage the line below. Ignore me. The answer is the primary header does not have a CCDNUM and I just wanted to be sure I wasn't reading a non-detector HDU (I don't know enough about decam to know if there are any other headers).

kfindeisen commented on python/astro_metadata_translator/translators/megaprime.py:
```diff
@@ -183,3 +185,62 @@ def to_observation_counter(self):
+            without already knowing what data is in the file. For many
+            instruments where the primary header is the only relevant
+            header, the primary header will be returned with no further
+            action.
```
Megaprime-specific wording?

kfindeisen commented on python/astro_metadata_translator/translators/megaprime.py:
```diff
@@ -183,3 +185,62 @@ def to_observation_counter(self):
+                    # handle the expected form.
+                    header = hdu.header
+
+                    # Astropy strips EXTNAME but translator needs it
+                    header["EXTNAME"] = hdu.name
```
Does this work correctly with Astropy > 4.2, which I assume does not strip it?

timj replied:
I think you are right that if my PR is accepted that the stripping should no longer be happening.

kfindeisen replied:
But does this code still work in that case, or does it need to be aware of the Astropy version it's running with? I don't have a clear picture of, e.g., whether this line would be a no-op in the absence of keyword stripping.

timj replied:
This codepath only triggers with a fixed astropy and I think I now stop the stripping. In that case it would be a no-op. I can remove the line for now since if my PR gets merged we will know if it's a problem straight away because the cfht metadata translation will fail in obs_chft test_ingest.py.

kfindeisen commented on python/astro_metadata_translator/translators/decam.py:
```diff
@@ -331,6 +331,12 @@ def read_all_headers(cls, filename, primary=None):
         # Circular dependency so must defer import
         from ..headers import merge_headers
 
+        # If this is not a FITS file return the primary if defined.
+        # It is likely that the metadata came from a test YAML file.
+        if not re.search(r"\.fit[s]?\b", filename) and primary is not None:
+            yield primary
+            return
+
```
I'm a bit concerned by how many things `read_all_headers` appears to be doing _other_ than reading all headers. Can you mention this special case in the base class documentation? It wouldn't even occur to me that it would be valid to call this method on a non-FITS file.

timj replied:
I've updated the documentation but now that I think about it it might be better to have `astrometadata translate` skip over this call completely if a test YAML file is encountered.

kfindeisen commented on python/astro_metadata_translator/cli/astrometadata.py:
```diff
@@ -26,10 +26,11 @@
                              " reading is attempted using the primary header. The primary header is"
-                             " always read and merged with this header.")
+                             " always read and merged with this header. Negative number indicates that"
+                             " the second header will be merged if the FITS file supports extended FITS.")
```
Perhaps mention that this is the default? I don't think Click output includes the default argument...

kfindeisen commented on python/astro_metadata_translator/translator.py:
```diff
@@ -977,7 +977,8 @@ def read_all_headers(cls, filename, primary=None):
 
-            yield read_basic_metadata_from_file(filename, 0)
+            # Merge primary and secondary header if they exist.
+            yield read_basic_metadata_from_file(filename, -1)
```
This contradicts the documented behavior, which suggests that if there are multiple headers you get multiple results. Does the description need to be updated to mention merging?

timj replied:
I have updated the documentation.

kfindeisen reviewed (changes requested):
I'm a bit worried about the behavior of the new `read_all_headers` method. It looks like it sometimes just returns the user-provided header, and sometimes not, sometimes considers alternatives, and sometimes not, and sometimes merges multiple headers, and sometimes not. This would be hard enough to reason about if it were documented, but since the docs just say it reads all headers and yields them one by one, any deviations from that behavior are going to be very surprising.

If this is a consequence of trying to unify too many different existing behaviors, then perhaps it would be less error-prone to have a custom ingest task for each instrument, where the user can at least know what to expect.

timj commented:
To reply to the global comment. read_all_headers is not meant to read every HDU. It's meant to read every header that is relevant for metadata translation. In the default case it yields back the primary header as given because it has no real knowledge of what should happen. If no primary header is given it attempts to read one -- merging the first two if it's a MEF. It can be reading a non-fits file because all the test code in astro_metadata_translator uses YAML files because they were small and easy to deal with. I could now add stripped files.

I'm not sure what a better name would be. "read_relevant_headers" is more accurate.

The idea is that the default is going to be mostly fine for most instruments that use one detector per file (and this is true for HSC, LATISS, LSSTCam etc).  DECam, CFHT and PhoSim subclass it since in the first two cases they have per-detector headers in one file and for PhoSim the primary header and the first amplifier header have different information that needs to be merged.  The subclass is allowed to do whatever is best for it but I tried to make the base class do the thing that works for most of our cameras.  The interface always yields one of more headers that are passed to metadata translation so it is all consistent.

timj commented:
There are two pull requests in astropy that will improve things:

* astropy/astropy#11396 should fix the problem with compressed_image EXTNAME masking the real EXTNAME.
* astropy/astropy#11404 should fix spurious warnings issued by the provenance code inserting HIERARCH headers into compressed headers.

kfindeisen commented:
> The subclass is allowed to do whatever is best for it but I tried to make the base class do the thing that works for most of our cameras. The interface always yields one of more headers that are passed to metadata translation so it is all consistent.

Strictly speaking, the subclass must do a subset of the base behavior, otherwise code that doesn't know the exact class it's dealing with (in particular, code in this package or in `obs_base`) may be working off false assumptions. It sounds from your other comments that you're relying on having a registry to always match up the right code with the right class, but that arrangement can be broken any time you add or modify a subclass.

In this case, I think the solution is to broaden the base class method's description to say that it returns one or more headers with enough information to process a file, without implying either that it returns all headers or that the headers it returns are necessarily those in the file (which covers merging and similar pre-processing). The formal return type is, as you say, not a problem.

timj commented:
I agree that the description is lacking and I will change it to make clear why returning the primary without reading anything else is reasonable.

I'm not sure I understand the comment about things breaking unexpectedly. The "registry" at issue here is the set of translator classes themselves and that's where the implementation is found. The system reads the primary (or first and second headers merged), asks which translator class to use and then asks that translator class if any more information should be read from the file to improve translation. I think that's all solid. If someone has an HSC file and asks for the DECam translator class to read more header information then they get what they get.

timj committed:
Rename read_all_headers and improve documentation for it

timj committed:
Add comment about CCDNUM

timj committed:
Remove EXTNAME setting for MegaPrime

I believe that in a situation where we have a ccdNN HDU,
the EXTNAME will not have been stripped.

timj committed:
Special case YAML metadata files

This removes the special casing in the DECam and MegaPrime
translators because now `astrometadata translate` decides
not to ask the translator class to read additional headers
if it knows it has a YAML file (which must be a test header).

timj commented:
@kfindeisen can you please do another review to see if you are happy with the current changes?

kfindeisen commented on python/astro_metadata_translator/file_helpers.py:
```diff
@@ -121,12 +122,13 @@ def read_basic_metadata_from_file(file, hdrnum, errstream=sys.stderr, can_raise=
         Header number to read. Only relevant for FITS. If greater than 1
-        it will be merged with the primary header.
+        it will be merged with the primary header. If negative the second
+        header will be merged with the primary header if a second is present.
```
This wording is a bit confusing, especially in the context of the previous sentence. Maybe say "If any negative number, the second header need not exist but will be merged with the primary header if it does."?

timj replied:
I've tried to rewrite.

kfindeisen commented on python/astro_metadata_translator/translator.py:
```diff
@@ -943,6 +946,61 @@ def to_observation_counter(self):
+
+        This method can optionally be given a header from the file.  This
+        header will generally be the primary header or a merge of the first
+        two headers.  In the base class implementation it is assumed that
+        this supplied header is the only useful header for metadata translation
+        and it will be returned unchanged if given. This can avoid
+        unnecesarily re-opening the file and re-reading the header when the
+        content is already known.
+
+        If no header is supplied, a header will be read from the supplied
+        file using `read_basic_metadata_from_file`, allowing it to merge
+        the primary and secondary header of a multi-extension FITS file.
```
I recommend moving the paragraph break to between "first two headers" and "In the base class implementation". Otherwise, it's not clear that the statement about `read_basic_metadata_from_file` applies only to this implementation and not to `determine_translatable_headers`. (I'm assuming, of course, that there is no requirement that subclasses call that particular function?)

timj replied:
Subclasses can do whatever they need to do to get a header from a file. I've added the line break.

kfindeisen commented on python/astro_metadata_translator/translator.py:
```diff
@@ -943,6 +946,61 @@ def to_observation_counter(self):
+        the primary and secondary header of a multi-extension FITS file.
+
+        Subclasses can return multiple headers and ignore the externally
+        supplied header.
```
Add that subclasses are also allowed to merge headers as appropriate for that instrument?

kfindeisen commented on python/astro_metadata_translator/translator.py:
```diff
@@ -943,6 +946,61 @@ def to_observation_counter(self):
+            without already knowing what data is in the file. For many
+            instruments where the primary header is the only relevant
+            header, the primary header will be returned with no further
+            action.
```
👍 , this is much clearer than the original wording.

kfindeisen commented on python/astro_metadata_translator/translator.py:
```diff
@@ -943,6 +946,61 @@ def to_observation_counter(self):
+        Yields
+        ------
+        headers : iterator of `dict`-like
+            Each relevant header in turn. Can be the supplied primary header.
```
"Can include"?

I wonder if this is a little too generic, since it makes it hard for the caller to know if the primary header is redundant with the return value. Does this also need to be special-cased by instrument?

timj replied:
I'll try to clarify. The method is only required to return a usable header for translation. How the method implements that is up to the method. If it wants to use the supplied header it can or else it can ignore it completely. I was trying not to repeat the general text from above.

kfindeisen replied:
I wasn't asking about whether it uses the supplied argument, but about whether `headers` includes the primary header at all (including a primary header it's read internally). From the DECam and Megacam examples it looks like you only want to return the primary header if there are no others, and if there are secondary headers then you never want to return the primary header separately?

timj replied:
I'm not understanding why the caller needs to know. Some instruments may return some information from the supplied header, other instruments may not. The only reason primary exists is to help out the general case by allowing the file open to be bypassed for instruments where the primary header is going to be enough for translation. MegaPrime will never look at the supplied header because by definition it's irrelevant. DECam will use it if given and merge the detector headers over the top. HSC can return the header immediately because it knows that there is only one header in an HSC file.

timj replied:
If primary=None the method still has to return the correct answer.

kfindeisen replied:
Perhaps say "Includes the primary header, if this header is needed for translation."? That makes it clear that the caller can just trust the output, instead of making the presence/absence of primary sound like an arbitrary choice.

(Of course, that would invalidate my requests for explicit mention of the primary header in the subclasses.)

timj replied:
The entire method assumes that each header returned by it should be directly usable as input to ObservationInfo constructor.  It shouldn't be trying to check if what was given as the header is identical to what is returned. Presence or absence of primary in returned header is instrument-specific and not something that caller needs to care about.

kfindeisen replied:
Yes, exactly! But "Can be the supplied primary header." implies otherwise.

timj replied:
Ah 😄 -- I'll remove that comment.

timj replied:
I had already removed that comment.

kfindeisen commented on python/astro_metadata_translator/translators/decam.py:
```diff
@@ -297,3 +298,65 @@ def fix_header(cls, header, instrument, obsid, filename=None):
+        Yields
+        ------
+        headers : iterator of `dict`-like
+            Each detector header in turn.
```
Explicitly state that this does not include an unmodified `primary`?

kfindeisen commented on python/astro_metadata_translator/translators/decam.py:
```diff
@@ -297,3 +298,65 @@ def fix_header(cls, header, instrument, obsid, filename=None):
+
+        Notes
+        -----
+        Each translator class can have code specifically tailored to its
```
I think this block might be redundant for subclass methods...

timj replied:
modified to say DECam.

kfindeisen commented on python/astro_metadata_translator/translators/megaprime.py:
```diff
@@ -183,3 +185,63 @@ def to_observation_counter(self):
+        MegaPrime files are multi-extension FITS with a primary header and
+        each detector stored in a subsequent extension.  MegaPrime uses
+        ``INHERIT=F`` therefore the primary header will always be ignored
+        if given.
```
Does "ignored" mean "not returned in `headers`"?

timj replied:
It means that whatever you supply as `primary` will make no difference to what is returned. INHERIT=F declares that each HDU has a standalone header that depends on no other header.

kfindeisen replied:
Then, as with the DECam translator, I suggest explicitly saying that the returned headers don't include the primary header.

timj replied:
To me "ignored" implies "will not be used". I have explicitly stated this now also in the yield line.

kfindeisen reviewed (approved):
This looks much better. I still have some questions about consistency across classes (in particular, the handling of `primary`), but nothing that would block merging.

timj committed:
Note the default hdrnum is negative for astrometadata translate

timj committed:
Make some improvements to documentation
